pkg/chunk/storage/factory.go

```diff
@@ -28,17 +28,28 @@ const (
 	StorageEngineTSDB   = "tsdb"
 )
 
+type indexClientFactories struct {
+	indexClientFactoryFunc IndexClientFactoryFunc
+	tableClientFactoryFunc TableClientFactoryFunc
+}
+
+// IndexClientFactoryFunc defines signature of function which creates chunk.IndexClient for managing index in index store
 type IndexClientFactoryFunc func() (chunk.IndexClient, error)
 
-var customIndexClients = map[string]IndexClientFactoryFunc{}
+// TableClientFactoryFunc defines signature of function which creates chunk.TableClient for managing tables in index store
+type TableClientFactoryFunc func() (chunk.TableClient, error)
 
-func RegisterIndexClient(name string, factory IndexClientFactoryFunc) {
-	customIndexClients[name] = factory
+var customIndexClients = map[string]indexClientFactories{}
+
+// RegisterIndexType is used for registering a custom index type.
+// When an index type is registered here with same name as existing types, the registered one takes the precedence.
+func RegisterIndexType(name string, indexClientFactory IndexClientFactoryFunc, tableClientFactory TableClientFactoryFunc) {
```

Comment:
As discussed offline, we agreed on naming this `RegisterIndexStore` and rename "index client" into "index store" across the file (as far as concerning the custom index store registration).